xdg-autostart-service: expand tilde in Exec lines

davidedmundson opened the pull request:
In typical desktop file parsing it is expected that "~" expands to a home directory.

Users may write an autostart file with "Exec=myCoolService ~/.someSpecialConfig" which worked before the systemd migration.

Downstream bug: https://bugs.kde.org/show_bug.cgi?id=455252#c18

yuwata commented on src/xdg-autostart-generator/xdg-autostart-service.c:
```diff
@@ -444,8 +450,11 @@ int xdg_autostart_format_exec_start(
                 percent = strreplace(raw, "%", "%%");
                 if (!percent)
                         return log_oom();
+                tilde_expanded = strreplace(percent, "~", home);
```
I do not think we should replace all `"~"`, e.g. `"foo~baz~aaa"`.
I think we should replace tilde only when the string starts with single tilde and remaining is empty or path:
- `"~"` -> `"/home/username"`
- `"~/path/to/file"` -> `"/home/username/path/to/file"`

yuwata commented on src/xdg-autostart-generator/xdg-autostart-service.c:
```diff
@@ -441,7 +442,22 @@ int xdg_autostart_format_exec_start(
 
-                free_and_replace(exec_split[n++], percent);
+                /*
+                 * Expand ~ if it comes as part of a path
```
Strictly speaking, 'part of a path' -> 'beginning of a path'

yuwata commented on src/xdg-autostart-generator/xdg-autostart-service.c:
```diff
@@ -441,7 +442,22 @@ int xdg_autostart_format_exec_start(
+                /*
+                 * Expand ~ if it comes as part of a path
+                 */
+                if (percent[0] == '~' && (strlen(percent) == 1 || percent[1] == '/')) {
```
Not a matter, but using `isempty()` and `path_is_absolute()` is better.
```
if (percent[0] == '~' && (isempty(percent + 1) || path_is_absolute(percent + 1))) {
```

yuwata commented on src/xdg-autostart-generator/xdg-autostart-service.c:
```diff
@@ -441,7 +442,22 @@ int xdg_autostart_format_exec_start(
+                 * Expand ~ if it comes as part of a path
+                 */
+                if (percent[0] == '~' && (strlen(percent) == 1 || percent[1] == '/')) {
+                        _cleanup_free_ char *home = NULL;
+                        r = get_home_dir(&home);
```
Please insert an empty line between variable declaration and code.

yuwata commented on src/xdg-autostart-generator/xdg-autostart-service.c:
```diff
@@ -441,7 +442,22 @@ int xdg_autostart_format_exec_start(
+                        if (!tilde_expanded)
+                                return log_oom();
+                        free_and_replace(exec_split[n++], tilde_expanded);
+                } else {
+                        free_and_replace(exec_split[n++], percent);
+                }
```
Please drop unnecessary `{}` for single line block.

yuwata reviewed (changes requested):
Looks mostly good. Several minor suggestions or coding style issues.

davidedmundson commented:
Thanks for the help, it's always tricky with coding styles and string functions when jumping into foreign projects.

davidedmundson commented:
>The test is still broken when running as part of TEST-02:

Right, the code is doing the right things. The test fails as I'm comparing to "getenv("HOME")" which apparently isn't always set. I've ported the test to use the same get_home_dir function

poettering commented on src/xdg-autostart-generator/xdg-autostart-service.c:
```diff
@@ -441,7 +442,22 @@ int xdg_autostart_format_exec_start(
+                        if (r < 0)
+                                return r;
+
+                        tilde_expanded = strjoin(home, &percent[1]);
```
path_join() sounds safer?

poettering commented:
Hmm, ~ is only allowed at the beginning of arguments?

AFAICS the .deskop spec doesn't specify this syntax. And intention to add it there?

poettering commented:
Did GNOME also support this ~ extension? Or was that a KDE thing?

poettering commented on src/xdg-autostart-generator/xdg-autostart-service.c:
```diff
@@ -441,7 +442,22 @@ int xdg_autostart_format_exec_start(
-                free_and_replace(exec_split[n++], percent);
+                /*
+                 * Expand ~ if it comes at the beginning of an argument to form a path
+                 */
```
Can you include in this comment that this is an extension to the .desktop spec?

benzea commented:
> Did GNOME also support this ~ extension? Or was that a KDE thing?

No, it does not look like GNOME ever supported this (i.e. neither the .desktop file code nor the used g_shell_parse_argv API does it).

davidedmundson commented:
Sorry for the delay, it's easy to miss comments in a merged PR.

I have addressed the technical parts at https://github.com/systemd/systemd/pull/24963

>Did GNOME also support this ~ extension? Or was that a KDE thing?

Apparently it's only a KDE thing. It's not explicitly supported in the spec to be expanded. But it is listed as saying it should be escaped if it is to be preserved. For that reason this patch also explicitly doesn't break the spec. 

Unfortunately in KDE we had quite a lot of flexible Exec parsing; which wasn't problematic at the time, but it's caused some fallout after we switched. Some we've sent back to the users, some I've tried to upstream if I thought it was more universal and not risky.

It's only at the start of args as the goal is to have a more shell like syntax.
